### ISSUE TITLE
`HttpReceiver#terminateResponse` may not be triggered so the request never ends

### ISSUE THREAD

**Issue by tangjiangling** [eclipse/jetty.project#8262]
**Jetty version(s)**
9.4.26.v20200117

**Java version/vendor** `(use: java -version)`
11.0.15+10

**OS type/version**
CentOS Linux release 7.4.1708

**Description**
We use jetty-client's asynchronous request method with [io.airlift.http.client.jetty.BufferingResponseListener](https://github.com/airlift/airlift/blob/master/http-client/src/main/java/io/airlift/http/client/jetty/BufferingResponseListener.java) to listen for the end of the request, and now we find that the request does not end when the request times out.

`HttpReceiver` DEBUG log:
```
2022-06-30T19:48:22.802+0800    DEBUG   http-client-exchange-scheduler-1        org.eclipse.jetty.client.HttpReceiver   Request/Response failed: Result[HttpRequest[GET /v1/task/20220630_194749_47322_wm993.2.8/results/3/0 HTTP/1.1]@260ff9bf > HttpResponse[null 0 null]@5b56baf0] java.util.concurrent.TimeoutException: Total timeout 10000 ms elapsed
2022-06-30T19:48:22.802+0800    DEBUG   http-client-exchange-scheduler-1        org.eclipse.jetty.client.HttpReceiver   Response abort HttpResponse[null 0 null]@486046b4 HttpExchange@219211c7 req=TERMINATED/null@null res=COMPLETED/java.util.concurrent.TimeoutException: Total timeout 10000 ms elapsed@757ae70e on HttpChannelOverHTTP@76763956(exchange=HttpExchange@219211c7 req=TERMINATED/null@null res=COMPLETED/java.util.concurrent.TimeoutException: Total timeout 10000 ms elapsed@757ae70e)[send=HttpSenderOverHTTP@2454bfb7(req=QUEUED,snd=COMPLETED,failure=null)[HttpGenerator@1946f356{s=START}],recv=HttpReceiverOverHTTP@5e3dc5a1(rsp=FAILURE,failure=java.util.concurrent.TimeoutException: Total timeout 10000 ms elapsed)[HttpParser{s=CLOSE,0 of -1}]]: java.util.concurrent.TimeoutException: Total timeout 10000 ms elapsed
2022-06-30T19:48:22.802+0800    DEBUG   http-client-exchange-scheduler-1        org.eclipse.jetty.client.HttpReceiver   Response complete HttpResponse[null 0 null]@486046b4
2022-06-30T19:48:22.802+0800    DEBUG   http-client-exchange-scheduler-1        org.eclipse.jetty.client.HttpReceiver   Request/Response failed: Result[HttpRequest[GET /v1/task/20220630_194809_47586_wm993.2.10/results/1/0 HTTP/1.1]@2ce12a1f > HttpResponse[null 0 null]@486046b4] java.util.concurrent.TimeoutException: Total timeout 10000 ms elapsed
2022-06-30T19:48:22.803+0800    DEBUG   http-client-exchange-scheduler-1        org.eclipse.jetty.client.HttpReceiver   Response abort HttpResponse[null 0 null]@667e9cd1 HttpExchange@57d3b5f0 req=TERMINATED/null@null res=COMPLETED/java.util.concurrent.TimeoutException: Total timeout 10000 ms elapsed@75336a93 on HttpChannelOverHTTP@5775cc1c(exchange=HttpExchange@57d3b5f0 req=TERMINATED/null@null res=COMPLETED/java.util.concurrent.TimeoutException: Total timeout 10000 ms elapsed@75336a93)[send=HttpSenderOverHTTP@3dfa27bf(req=QUEUED,snd=COMPLETED,failure=null)[HttpGenerator@422e12c4{s=START}],recv=HttpReceiverOverHTTP@3bc233b1(rsp=FAILURE,failure=java.util.concurrent.TimeoutException: Total timeout 10000 ms elapsed)[HttpParser{s=CLOSE,0 of -1}]]: java.util.concurrent.TimeoutException: Total timeout 10000 ms elapsed
2022-06-30T19:48:22.803+0800    DEBUG   http-client-exchange-scheduler-1        org.eclipse.jetty.client.HttpReceiver   Response complete HttpResponse[null 0 null]@667e9cd1
2022-06-30T19:48:22.803+0800    DEBUG   http-client-exchange-scheduler-1        org.eclipse.jetty.client.HttpReceiver   Request/Response failed: Result[HttpRequest[GET /v1/task/20220630_194756_47402_wm993.2.10/results/14/0 HTTP/1.1]@50a717f8 > HttpResponse[null 0 null]@667e9cd1] java.util.concurrent.TimeoutException: Total timeout 10000 ms elapsed
2022-06-30T19:48:22.803+0800    DEBUG   http-client-exchange-scheduler-1        org.eclipse.jetty.client.HttpReceiver   Response abort HttpResponse[null 0 null]@422d8129 HttpExchange@89bd0a9 req=TERMINATED/null@null res=COMPLETED/java.util.concurrent.TimeoutException: Total timeout 10000 ms elapsed@669244db on HttpChannelOverHTTP@62e7f8f1(exchange=HttpExchange@89bd0a9 req=TERMINATED/null@null res=COMPLETED/java.util.concurrent.TimeoutException: Total timeout 10000 ms elapsed@669244db)[send=HttpSenderOverHTTP@1f8d0ae5(req=QUEUED,snd=COMPLETED,failure=null)[HttpGenerator@6f3020d{s=START}],recv=HttpReceiverOverHTTP@5b8e5252(rsp=FAILURE,failure=java.util.concurrent.TimeoutException: Total timeout 10000 ms elapsed)[HttpParser{s=CLOSE,0 of -1}]]: java.util.concurrent.TimeoutException: Total timeout 10000 ms elapsed
2022-06-30T19:48:22.803+0800    DEBUG   http-client-exchange-scheduler-1        org.eclipse.jetty.client.HttpReceiver   Response complete HttpResponse[null 0 null]@422d8129
2022-06-30T19:48:23.014+0800    DEBUG   http-client-exchange-scheduler-1        org.eclipse.jetty.client.HttpReceiver   Response abort HttpResponse[HTTP/1.1 204 No Content]@41c584d4 HttpExchange@1a00d7ed req=TERMINATED/null@null res=COMPLETED/java.util.concurrent.TimeoutException: Total timeout 10000 ms elapsed@75ccd6ca on HttpChannelOverHTTP@1c3d6b50(exchange=HttpExchange@1a00d7ed req=TERMINATED/null@null res=COMPLETED/java.util.concurrent.TimeoutException: Total timeout 10000 ms elapsed@75ccd6ca)[send=HttpSenderOverHTTP@22e2fc65(req=QUEUED,snd=COMPLETED,failure=null)[HttpGenerator@416fdb69{s=START}],recv=HttpReceiverOverHTTP@36ff5fff(rsp=FAILURE,failure=java.util.concurrent.TimeoutException: Total timeout 10000 ms elapsed)[HttpParser{s=CLOSE,0 of -1}]]: java.util.concurrent.TimeoutException: Total timeout 10000 ms elapsed
2022-06-30T19:48:23.014+0800    DEBUG   http-client-exchange-scheduler-1        org.eclipse.jetty.client.HttpReceiver   Concurrent failure: response termination skipped, performed by helpers
```

We found this debug log: `Concurrent failure: response termination skipped, performed by helpers`.

It seems that `HttpReceiver#abort` doesn't trigger `HttpReceiver#terminateResponse`, and thus doesn't trigger `Response.CompleteListener#onComplete`, eventually the request won't end either.

We can refer to: https://github.com/eclipse/jetty.project/blob/jetty-9.4.26.v20200117/jetty-client/src/main/java/org/eclipse/jetty/client/HttpReceiver.java#L613-L625

**Comment by gregw** [comment]
Can you attach a code snippet to show how you are setting up and using the client

**Comment by tangjiangling** [comment]
> Can you attach a code snippet to show how you are setting up and using the client

You can refer to https://github.com/airlift/airlift/blob/0.192/http-client/src/main/java/io/airlift/http/client/jetty/JettyHttpClient.java#L142-L371

Simplify it:

For `HttpClient`:
```
... // some codes
HttpClient httpClient = new HttpClient(transport, sslContextFactory);

httpClient.setRequestBufferSize(4KB);
httpClient.setResponseBufferSize(16KB);

httpClient.setMaxConnectionsPerDestination(20);
httpClient.setMaxRequestsQueuedPerDestination(1024);

httpClient.setIdleTimeout(30S);
httpClient.setConnectTimeout(5S);
httpClient.setAddressResolutionTimeout(5S);
```

For `HttpRequest`:
```
... // some codes
jettyRequest.timeout(10000, MILLISECONDS); // 10S
jettyRequest.queueTimeout(300000, MILLISECONDS); // 5min
jettyRequest.idleTimeout(30000, MILLISECONDS); //30S
```

For execution, you can refer to https://github.com/airlift/airlift/blob/0.192/http-client/src/main/java/io/airlift/http/client/jetty/JettyHttpClient.java#L550

**Comment by gregw** [comment]
Thanks,

so the request is built [here](https://github.com/airlift/airlift/blob/0.192/http-client/src/main/java/io/airlift/http/client/jetty/JettyHttpClient.java#L599-L644).

I'm assuming because it is a GET, then `finalRequest.getBodyGenerator();` is null.  
Do you know the values of `finalRequest.isFollowRedirects()` and `finalRequest.isPreserveAuthorizationOnRedirect()`?

**Comment by tangjiangling** [comment]
> I'm assuming because it is a GET

Yes.

> Do you know the values of finalRequest.isFollowRedirects() and finalRequest.isPreserveAuthorizationOnRedirect()?

- `finalRequest.isFollowRedirects()` = `true`
- `finalRequest.isPreserveAuthorizationOnRedirect()` = `false`

**Comment by tangjiangling** [comment]
This is the logic of the server-side processing (which may help you a bit ):
https://github.com/trinodb/trino/blob/330/presto-main/src/main/java/io/prestosql/server/TaskResource.java#L242
(There will be some waiting time during processing)

**Comment by gregw** [comment]
OK, I think I have enough info to try to reproduce now... but I may not have enough time for a few days.
Stand by...

**Comment by tangjiangling** [comment]
Never mind. (We temporarily let `listener` implement the `Response.FailureListener#onFailure` interface to ensure that the request exits properly)

**Comment by gregw** [comment]
Well we will leave open and investigate when we have time, as non completion can be a bad bug.

**Comment by sbordet** [comment]
@tangjiangling can you confirm that `onComplete()` does not get called when you see `Concurrent failure: response termination skipped, performed by helpers`?

I ask because we are able to reproduce the log line, but `onComplete()` is always called.

**Comment by sbordet** [comment]
@tangjiangling can you please update to Jetty 9.4.48, as 9.4.26 is really out of date.

**Comment by tangjiangling** [comment]
> can you confirm that onComplete() does not get called when you see Concurrent failure: response termination skipped, performed by helpers?

Yep, I confirmed.

You can try manually setting [terminate](https://github.com/eclipse/jetty.project/blob/jetty-9.4.26.v20200117/jetty-client/src/main/java/org/eclipse/jetty/client/HttpReceiver.java#L613-L619) to `false` in DEBUG mode and you will see that `onComplete` will never be called.

> can you please update to Jetty 9.4.48, as 9.4.26 is really out of date.

Sure, but I'm curious to know what causes this issue.

**Comment by sbordet** [comment]
Do you have DEBUG logs that show the issue? I cannot reproduce it.

The `terminate` boolean may be `false` and when it is other threads will terminate the response and call `onComplete()`.

**Comment by tangjiangling** [comment]
> Do you have DEBUG logs that show the issue?

No, I didn't.

> You can try manually setting [terminate](https://github.com/eclipse/jetty.project/blob/jetty-9.4.26.v20200117/jetty-client/src/main/java/org/eclipse/jetty/client/HttpReceiver.java#L613-L619) to false in DEBUG mode and you will see that onComplete will never be called.

I confirmed this by `debug` the jetty unit test (`HttpReceiverOverHTTPTest#testReceiveResponseContent`) and manually setting `terminate` to `false`.

**Comment by sbordet** [comment]
Ok let me rephrase.
You posted DEBUG logs in your first message, so I thought you had DEBUG logs?

About changing the value of the `terminate` variable with the debugger, I don't see why you're doing it?
If it was `true` and you changed it to `false` obviously the response won't terminate.

But when `terminate=false` happens without external intervention through a debugger, it means that another thread is doing the termination.

**Comment by tangjiangling** [comment]
> You posted DEBUG logs in your first message, so I thought you had DEBUG logs?

Yep.(The full debug log is in the `Description` of the issue)

> About changing the value of the terminate variable with the debugger, I don't see why you're doing it?

Since you said you could reproduce the exception log (`Concurrent failure: response termination skipped, performed by helpers `), I intentionally made this happen to verify that `onComplete` would be executed.

**Comment by sbordet** [comment]
Sorry I don't understand.

> (The full debug log is in the Description of the issue)

Does not look the full DEBUG log to me because it misses a large number of lines that would be printed for each response. 
For example, this is what I get grepping for a single response object:

```
2022-07-07 13:02:41.248:DEBUG:oejc.HttpDestination:ForkJoinPool.commonPool-worker-12: Processing exchange HttpExchange@4fcd6329{req=HttpRequest[GET /thread_2_request_2 HTTP/1.1]@2202225b[PENDING/null] res=HttpResponse[null 0 null]@1a831ca4[PENDING/null]} on HttpConnectionOverHTTP@23ea723a::SocketChannelEndPoint@6ab47a25[{l=/127.0.0.1:42710,r=localhost/127.0.0.1:45929,OPEN,fill=FI,flush=-,to=0/250}{io=1/1,kio=1,kro=8}]->[HttpConnectionOverHTTP@23ea723a(l:/127.0.0.1:42710 <-> r:localhost/127.0.0.1:45929,closed=false)=>HttpChannelOverHTTP@15989861(exchange=null)[send=HttpSenderOverHTTP@5231360f(req=QUEUED,failure=null)[HttpGenerator@223ae5d1{s=START}],recv=HttpReceiverOverHTTP@6b96d6e1(rsp=IDLE,failure=null)[HttpParser{s=START,0 of -1}]]] of HttpDestination[Origin@ddac247a[http://localhost:45929,tag=null,protocol=Protocol@7f4624a[proto=[http/1.1],nego=false]]]@41813449,queue=22,pool=DuplexConnectionPool@4678a2eb[c=24/47/64,a=23,i=0,q=22]
2022-07-07 13:02:41.248:DEBUG:oejc.HttpChannel:ForkJoinPool.commonPool-worker-12: HttpExchange@4fcd6329{req=HttpRequest[GET /thread_2_request_2 HTTP/1.1]@2202225b[PENDING/null] res=HttpResponse[null 0 null]@1a831ca4[PENDING/null]} associated true to HttpChannelOverHTTP@15989861(exchange=HttpExchange@4fcd6329{req=HttpRequest[GET /thread_2_request_2 HTTP/1.1]@2202225b[PENDING/null] res=HttpResponse[null 0 null]@1a831ca4[PENDING/null]})[send=HttpSenderOverHTTP@5231360f(req=QUEUED,failure=null)[HttpGenerator@223ae5d1{s=START}],recv=HttpReceiverOverHTTP@6b96d6e1(rsp=IDLE,failure=null)[HttpParser{s=START,0 of -1}]]
2022-07-07 13:02:41.249:DEBUG:oejc.HttpExchange:ForkJoinPool.commonPool-worker-12: Terminated request for HttpExchange@4fcd6329{req=HttpRequest[GET /thread_2_request_2 HTTP/1.1]@2202225b[TERMINATED/null] res=HttpResponse[null 0 null]@1a831ca4[PENDING/null]}, result: null
2022-07-07 13:02:41.510:DEBUG:oejc.HttpConnection:HttpClient@27305e6-scheduler-1: Idle timeout 260/250ms - HttpConnectionOverHTTP@23ea723a::SocketChannelEndPoint@6ab47a25[{l=/127.0.0.1:42710,r=localhost/127.0.0.1:45929,OPEN,fill=-,flush=-,to=261/250}{io=0/0,kio=0,kro=1}]->[HttpConnectionOverHTTP@23ea723a(l:/127.0.0.1:42710 <-> r:localhost/127.0.0.1:45929,closed=false)=>HttpChannelOverHTTP@15989861(exchange=HttpExchange@4fcd6329{req=HttpRequest[GET /thread_2_request_2 HTTP/1.1]@2202225b[TERMINATED/null] res=HttpResponse[null 0 null]@1a831ca4[PENDING/null]})[send=HttpSenderOverHTTP@5231360f(req=QUEUED,failure=null)[HttpGenerator@223ae5d1{s=START}],recv=HttpReceiverOverHTTP@6b96d6e1(rsp=IDLE,failure=null)[HttpParser{s=START,0 of -1}]]]
2022-07-07 13:02:41.510:DEBUG:oejc.AbstractConnectionPool:HttpClient@27305e6-scheduler-1: Removed (true) MultiEntry@dcfda20{CLOSED,usage=0,multiplex=0,pooled=HttpConnectionOverHTTP@23ea723a::SocketChannelEndPoint@6ab47a25[{l=/127.0.0.1:42710,r=localhost/127.0.0.1:45929,OPEN,fill=-,flush=-,to=0/250}{io=0/0,kio=0,kro=1}]->[HttpConnectionOverHTTP@23ea723a(l:/127.0.0.1:42710 <-> r:localhost/127.0.0.1:45929,closed=true)=>HttpChannelOverHTTP@15989861(exchange=HttpExchange@4fcd6329{req=HttpRequest[GET /thread_2_request_2 HTTP/1.1]@2202225b[TERMINATED/null] res=HttpResponse[null 0 null]@1a831ca4[PENDING/null]})[send=HttpSenderOverHTTP@5231360f(req=QUEUED,failure=null)[HttpGenerator@223ae5d1{s=START}],recv=HttpReceiverOverHTTP@6b96d6e1(rsp=IDLE,failure=null)[HttpParser{s=START,0 of -1}]]]} Pool@14edf7d6[inUse=63,size=63,max=64,closed=false]
2022-07-07 13:02:41.510:DEBUG:oejch.HttpReceiverOverHTTP:HttpClient@27305e6-2042: Read 108 bytes in RetainableByteBuffer@5b5d60{DirectByteBuffer@74348ea8[p=0,l=108,c=16384,r=108]={<<<HTTP/1.1 200 OK\r\nDate: Th...y(10.0.12-SNAPSHOT)\r\n\r\n>>>\x00\x00\x00\x00\x00\x00\x00\x00\x00...\x00\x00\x00\x00\x00\x00\x00},r=1} from SocketChannelEndPoint@6ab47a25[{l=/127.0.0.1:42710,r=localhost/127.0.0.1:45929,OPEN,fill=-,flush=-,to=0/250}{io=0/0,kio=0,kro=1}]->[HttpConnectionOverHTTP@23ea723a(l:/127.0.0.1:42710 <-> r:localhost/127.0.0.1:45929,closed=false)=>HttpChannelOverHTTP@15989861(exchange=HttpExchange@4fcd6329{req=HttpRequest[GET /thread_2_request_2 HTTP/1.1]@2202225b[TERMINATED/null] res=HttpResponse[null 0 null]@1a831ca4[PENDING/null]})[send=HttpSenderOverHTTP@5231360f(req=QUEUED,failure=null)[HttpGenerator@223ae5d1{s=START}],recv=HttpReceiverOverHTTP@6b96d6e1(rsp=IDLE,failure=null)[HttpParser{s=START,0 of -1}]]]
2022-07-07 13:02:41.510:DEBUG:oejc.HttpExchange:HttpClient@27305e6-scheduler-1: Failed HttpExchange@4fcd6329{req=HttpRequest[GET /thread_2_request_2 HTTP/1.1]@2202225b[TERMINATED/null] res=HttpResponse[HTTP/1.1 200 OK]@1a831ca4[COMPLETED/java.util.concurrent.TimeoutException: Idle timeout 250 ms]}: req=false/rsp=true {}
2022-07-07 13:02:41.511:DEBUG:oejc.HttpReceiver:HttpClient@27305e6-scheduler-1: Response abort HttpResponse[HTTP/1.1 200 OK]@1a831ca4 HttpExchange@4fcd6329{req=HttpRequest[GET /thread_2_request_2 HTTP/1.1]@2202225b[TERMINATED/null] res=HttpResponse[HTTP/1.1 200 OK]@1a831ca4[COMPLETED/java.util.concurrent.TimeoutException: Idle timeout 250 ms]} on HttpChannelOverHTTP@15989861(exchange=HttpExchange@4fcd6329{req=HttpRequest[GET /thread_2_request_2 HTTP/1.1]@2202225b[TERMINATED/null] res=HttpResponse[HTTP/1.1 200 OK]@1a831ca4[COMPLETED/java.util.concurrent.TimeoutException: Idle timeout 250 ms]})[send=HttpSenderOverHTTP@5231360f(req=QUEUED,failure=null)[HttpGenerator@223ae5d1{s=START}],recv=HttpReceiverOverHTTP@6b96d6e1(rsp=FAILURE,failure=java.util.concurrent.TimeoutException: Idle timeout 250 ms)[HttpParser{s=HEADER,0 of -1}]]: {}
2022-07-07 13:02:41.511:DEBUG:oejc.HttpReceiver:HttpClient@27305e6-2042: Response begin HttpResponse[HTTP/1.1 200 OK]@1a831ca4
2022-07-07 13:02:41.513:DEBUG:oejc.HttpExchange:HttpClient@27305e6-2042: Terminated response for HttpExchange@4fcd6329{req=HttpRequest[GET /thread_2_request_2 HTTP/1.1]@2202225b[TERMINATED/null] res=HttpResponse[HTTP/1.1 200 OK]@1a831ca4[TERMINATED/java.util.concurrent.TimeoutException: Idle timeout 250 ms]}, result: Result[HttpRequest[GET /thread_2_request_2 HTTP/1.1]@2202225b > HttpResponse[HTTP/1.1 200 OK]@1a831ca4] java.util.concurrent.TimeoutException: Idle timeout 250 ms
2022-07-07 13:02:41.513:DEBUG:oejc.HttpReceiver:HttpClient@27305e6-2042: Response complete HttpResponse[HTTP/1.1 200 OK]@1a831ca4, result: Result[HttpRequest[GET /thread_2_request_2 HTTP/1.1]@2202225b > HttpResponse[HTTP/1.1 200 OK]@1a831ca4] java.util.concurrent.TimeoutException: Idle timeout 250 ms
2022-07-07 13:02:41.513:DEBUG:oejc.HttpChannel:HttpClient@27305e6-2042: HttpExchange@4fcd6329{req=HttpRequest[GET /thread_2_request_2 HTTP/1.1]@2202225b[TERMINATED/null] res=HttpResponse[HTTP/1.1 200 OK]@1a831ca4[TERMINATED/java.util.concurrent.TimeoutException: Idle timeout 250 ms]} disassociated true from HttpChannelOverHTTP@15989861(exchange=null)[send=HttpSenderOverHTTP@5231360f(req=QUEUED,failure=null)[HttpGenerator@223ae5d1{s=START}],recv=HttpReceiverOverHTTP@6b96d6e1(rsp=FAILURE,failure=java.util.concurrent.TimeoutException: Idle timeout 250 ms)[HttpParser{s=HEADER,0 of -1}]]
2022-07-07 13:02:41.514:DEBUG:oejc.HttpReceiver:HttpClient@27305e6-2042: Request/Response failed: Result[HttpRequest[GET /thread_2_request_2 HTTP/1.1]@2202225b > HttpResponse[HTTP/1.1 200 OK]@1a831ca4] java.util.concurrent.TimeoutException: Idle timeout 250 ms, notifying [org.eclipse.jetty.client.HttpClientFailureTest$$Lambda$456/0x00000007c01af0a0@22272b3f]
2022-07-07 13:02:41.513:DEBUG:oejc.HttpReceiver:HttpClient@27305e6-scheduler-1: Concurrent failure: response termination skipped, performed by helpers HttpResponse[HTTP/1.1 200 OK]@1a831ca4 HttpExchange@4fcd6329{req=HttpRequest[GET /thread_2_request_2 HTTP/1.1]@2202225b[TERMINATED/null] res=HttpResponse[HTTP/1.1 200 OK]@1a831ca4[COMPLETED/java.util.concurrent.TimeoutException: Idle timeout 250 ms]} on HttpChannelOverHTTP@15989861(exchange=HttpExchange@4fcd6329{req=HttpRequest[GET /thread_2_request_2 HTTP/1.1]@2202225b[TERMINATED/null] res=HttpResponse[HTTP/1.1 200 OK]@1a831ca4[COMPLETED/java.util.concurrent.TimeoutException: Idle timeout 250 ms]})[send=HttpSenderOverHTTP@5231360f(req=QUEUED,failure=null)[HttpGenerator@223ae5d1{s=START}],recv=HttpReceiverOverHTTP@6b96d6e1(rsp=FAILURE,failure=java.util.concurrent.TimeoutException: Idle timeout 250 ms)[HttpParser{s=HEADER,0 of -1}]]
2022-07-07 13:02:43.585:DEBUG:oejc.HttpExchange:HttpClient@27305e6-scheduler-1: Aborted (false) while active HttpExchange@4fcd6329{req=HttpRequest[GET /thread_2_request_2 HTTP/1.1]@2202225b[TERMINATED/null] res=HttpResponse[HTTP/1.1 200 OK]@1a831ca4[TERMINATED/java.util.concurrent.TimeoutException: Idle timeout 250 ms]}: {}
```

You can see from the logs above that I can reproduce the log line "... performed by helpers" in the sense that I could write a test case that caused that particular race condition to happen, but the response was always terminated.

I don't understand if you are getting this problem in real world, or only by manually triggering it with the debugger?

If you get it in the real world, I need the full DEBUG logs.

**Comment by tangjiangling** [comment]
@sbordet I only enabled debug log for `HttpReceiver`, so only `HttpReceiver` related logs.

> I don't understand if you are getting this problem in real world, or only by manually triggering it with the debugger?
   If you get it in the real world, I need the full DEBUG logs.

I'll try to reproduce the problem, and if there are new progress I'll sync it here. (This issue is real in our production environment)

**Comment by tangjiangling** [comment]
@gregw @sbordet Thanks for your help, let me close this issue, as the problem is not reproducible after using `9.4.48` version.
